### PR TITLE
add set-pri() as a rewrite operation

### LINKED ISSUE #3541: Parse only PRI
## Description of the problem
In certain cases where the source is more invalid than valid and the syslog parser would rightfully mangle the event I don't call the syslog-ng parser instead I parse using patterndb or other features of syslog-ng.

For example
<134>1 1111611101486.978292274 CLL_FWL flows src=xxx.xxx.xxx.xxx dst=xxx.xxx.xxx.xxx protocol=tcp sport=xxxx dport=xxxx pattern: 1 all

## Proposed solution
Provide a parser or re-write function that would take "134" as a parameter set PRI, SEVERITY, and FACILITY

## Alternatives
Use Python to calculate facility and severity then call set-severity and set-facility re-write but this is more steps and python is slow for this use

### PR DESCRIPTION
**Pull request by bazsi**
This branch adds a new rewrite operation, set-pri() that allows one to set the pri value of the message (e.g. facility * 8 + severity) based on a template string.

The template string needs to expand to a number between in the interval [0, 1023] inclusive, that permits up to 127 facilities.

The format accepted by set-pri() is a simple number, e.g. if you extract this value from a syslog header, the brackets "<" and ">" needs to be removed first.

The format may be extended in the future if use-cases arise, I was thinking of allowing the brackets to be present, or the use of the dot separated, string based format, such as "mail.info".

Fixes #3541

### CODE REVIEW

**Comment by mbonsack** [comment]
Allowing the brackets to be present would be helpful, but not critical.  Thanks!

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/rewrite/rewrite-set-pri.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  LogRewrite super;
+  LogTemplate *pri;
+};
+
+static gint
+_convert_pri(GString *pri_text)
+{
+  char *endptr;
+  long int pri = strtol(pri_text->str, &endptr, 10);
+  if (0 == endptr)
```
Comment (alltilla) [inline lib/rewrite/rewrite-set-pri.c:44]: Should be `NULL`, right?

File: lib/rewrite/tests/test_set_pri.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+Test(set_pri, numeric)
+{
+  _perform_set_pri(_create_template("7"), msg);
+  cr_assert(_msg_pri_equals(msg, LOG_KERN | LOG_DEBUG));
+
+  _perform_set_pri(_create_template("189"), msg);
+  cr_assert(_msg_pri_equals(msg, LOG_LOCAL7 | LOG_NOTICE));
+
+  _perform_set_pri(_create_template("$PRI"), msg);
+  cr_assert(_msg_pri_equals(msg, 189));
```
Comment (alltilla) [inline lib/rewrite/tests/test_set_pri.c:84]: Removing `_perform_set_pri(_create_template("$PRI"), msg);` still produces green tests, so this test could hide problems, where the template resolution does not affect the previous `PRI` value.

I would rather test the template resolution with a different macro and different value, like:
```
  log_msg_set_value_by_name(msg, "MSG", "137", -1);
  _perform_set_pri(_create_template("$MSG"), msg);
  cr_assert(_msg_pri_equals(msg, 137));
```

**Review by alltilla — COMMENTED** [review]
I have found these 2 minor problems, otherwise it looks good to me!

File: lib/rewrite/rewrite-set-pri.c (hunk 1 above)
Comment (bazsi) [reply lib/rewrite/rewrite-set-pri.c]: fixed.

File: lib/rewrite/tests/test_set_pri.c (hunk 2 above)
Comment (bazsi) [reply lib/rewrite/tests/test_set_pri.c]: fixed this as well, and added a few more testcases to cover more invalid cases.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by alltilla — APPROVED** [review]

**Review by MrAnno — APPROVED** [review]

**Pull request merged**